src/display.rs

```diff
@@ -78,46 +77,89 @@ impl Display for InlineTable {
     }
 }
 
-struct TableFormatState<'t> {
-    path: RefCell<Vec<&'t str>>,
-    table: Cell<&'t Table>,
-}
-
-impl<'a> Display for TableFormatState<'a> {
-    fn fmt(&self, f: &mut Formatter) -> Result {
-        let table = self.table.get();
-
-        for kv in table.items.values() {
-            if let Item::Value(ref value) = kv.value {
-                writeln!(f, "{}={}", kv.key, value)?;
+impl Table {
+    fn visit_nested_tables_in_original_order<'t, F>(
+        &'t self,
+        path: &mut Vec<&'t str>,
+        is_array_of_tables: bool,
+        callback: &mut F,
+    ) -> Result
+    where
+        F: FnMut(&Table, &Vec<&'t str>, bool) -> Result,
+    {
+        let mut current_position = 0usize;
+        let mut last_position = 0usize;
+        let mut max_position = 0usize;
+        let mut should_print = true;
+        loop {
+            self.visit_nested_tables(path, is_array_of_tables, &mut |t, path, is_array| {
+                match &t.position {
+                    Some(ref pos) => {
+                        if *pos > max_position {
+                            max_position = *pos;
+                        }
+                        if *pos == current_position {
+                            // if this table is from from the position we're
+                            // looking for, print it.
+                            should_print = true;
+                            current_position += 1;
+                        } else {
+                            // if this table isn't from the position we're
+                            // looking for, skip it.
+                            should_print = false;
+                        }
+                    }
+                    // If this table doesn't have a position then it was
+                    // probably made programmatically, so put it after wherever
+                    // it happens to be after in the tree. We only want to print
+                    // it once though, so rely on should_print from the previous
+                    // table to tell you whether you need to print this one.
+                    None => (),
```

Comment:
Also, if you parse two .toml files and try to merge them, you will have duplicate position integers, and when you print them things will get skipped.